### DISCUSSION TITLE
how to determine the current width of a node in Groovy

### DISCUSSION THREAD

**Discussion by nnako** [freeplane/freeplane#650] [Help] [answered]
Hi,

a couple of years ago, I created a script which made it possible to automatically detect the node widths of selected nodes in order to identify the maximum width. here is an example script:

```groovy
import org.freeplane.view.swing.map.NodeView

// some initial values

def maxwidth = 0
def width = 0

// take each selected node

selecteds.each{

    // switch off any width limitation setting for
    // the particular node in order to show its real
    // width

    it.style.setMinNodeWidth( -1 )
    it.style.setMaxNodeWidth( -1 )

    // now, get current width of the node

    nodeView = it.delegate.viewers.find() {it instanceof NodeView}
    width = nodeView?.mainView?.width

    // update "maxwidth" if current width is larger

    if( width > maxwidth ){

        // update maximum width variable
        // with current node width

        maxwidth = width
    }
}
c.statusInfo = maxwidth.toString()
```

now, the part which does not seem to work properly is the calculation of `width` variable in the middle of the code. the return value seems to jump between values and seems to be about half the dimension of the parameters which can be put into the format menu for `minNodeWidth` and `maxNodeWidth`.

the value retrieved by the two code lines (determination of `nodeView` and `width`) seems to be dependent on the current zoom level, and maybe other parameters?

any hints, how to determine the current (limitation-free) width of a node using Groovy?

**Comment by nnako** [comment, marked as answer]
just an addition:

```groovy

    ...

    // switch off any width limitation setting for
    // the particular node in order to show its real
    // width

    it.style.setMinNodeWidth( -1 )
    it.style.setMaxNodeWidth( -1 )

    // now, get current width of the node

    nodeView = it.delegate.viewers.find() {it instanceof NodeView}
    width = nodeView?.mainView?.width

```

by trying to switch off all existing limitations, I want to ensure that I reset the node's `min` and `max` view settings **before** the width is determined by the next following two lines, but it seems as if the limitations are still active when reading the current node width. so, I just get the min limitation value instead of the pure node width.

strange...

**Reply by macmarrum** [reply]
I can see that `node.style.minNodeWidth = -1` is reflected in `node.delegate.viewers.find {it instanceof NodeView}?.mainView?.width / c.zoom` but only after an independent run (another script execution)

**Reply by nnako** [reply]
@macmarrum, yes, that seems to be the problem. the setting of `minNodeWidth` does not seem to be activated on the code level until the next script start. I will try to re-order my script and take this into account. thanks.
